### PR TITLE
Try to fix data race in `WriteBufferFromHTTPServerResponse`

### LINKED ISSUE #69520: Data race in `WriteBufferFromHTTPServerResponse`
https://s3.amazonaws.com/clickhouse-test-reports/69332/55d6672bc43579dad3edaa1e8c5d595025d50fc8/stateless_tests__tsan__[3_4]/stderr.log
 
This function can be called while the query is running and writing to the output buffer:
https://github.com/ClickHouse/ClickHouse/blob/master/src/Server/HTTP/WriteBufferFromHTTPServerResponse.cpp#L21
It's called in `onProgress` which acquires the mutex but `ParallelFormattingOutputFormat` ends up calling `offset` directly on the write buffer so no mutex is acquired.

### PR DESCRIPTION
**Pull request by thevar1able**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Fixes https://github.com/ClickHouse/ClickHouse/issues/69520

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit e3dd7b6668caa16d98be0ac1b570dd556511f6d3 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/69601/e3dd7b6668caa16d98be0ac1b570dd556511f6d3/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/2fce90ab76b851198b39041c9e7c4b03ad601ff7/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by yakov-olkhovskiy** [comment]
besides bug in `WriteBufferFromHTTPServerResponse` where access to buffer (and all corresponding variables) should be locked by existing mutex, the race condition in the example is a different story and an awful one - it's consequence of having intrusive buffers - compression wrapper is not aware of any concurrency of writing into underlying buffer and access its guts directly... apparently we need to make compression wrappers aware of need of locking underlying buffer too...

**Comment by antonio2368** [comment]
>  it's consequence of having intrusive buffers - compression wrapper is not aware of any concurrency of writing into underlying buffer and access its guts directly... apparently we need to make compression wrappers aware of need of locking underlying buffer too...

I.e. our buffers are not thread-safe and making them so could introduce unwanted complexity.
I would prefer keeping buffers simple like they are now especially if the problem can be solved in a different way.
IIUC here we protect with mutex writing to the socket and not writing to the buffer itself so we just need to do all the buffer modification (e.g. `setChunked`) before we send it further.

**Comment by yakov-olkhovskiy** [comment]
@antonio2368 @thevar1able per my assessment `setChunked` indeed can be called on `WriteBufferFromHTTPServerResponse` constructor, so we will not touch buffer concurrently from this point. Regarding protection - we only need to protect state around `headers_started_sending` and `headers_finished_sending` - we are sending HTTP header directly without using the buffer, so the idea here that we can't send anything header-related once body started.
I think I know what to do, @thevar1able are you ok if I take this issue?

**Comment by thevar1able** [comment]
@yakov-olkhovskiy sure, go ahead and open a PR

File: src/Server/StaticRequestHandler.cpp [hunk 1]
```diff
@@ -90,15 +90,15 @@ static inline void trySendExceptionToClient(
...
     auto out = responseWriteBuffer(request, response);
 
     try
     {
-        applyHTTPResponseHeaders(response, http_response_headers_override);
-
-        if (request.getVersion() == Poco::Net::HTTPServerRequest::HTTP_1_1)
-            response.setChunkedTransferEncoding(true);
-
         /// Workaround. Poco does not detect 411 Length Required case.
```
Comment (yakov-olkhovskiy) [inline src/Server/StaticRequestHandler.cpp:102]: oh, good catch... do we have other places like that?

**Comment by yakov-olkhovskiy** [comment]
...also I'm not sure regarding `headers_started_sending` and `headers_finished_sending` - it seems they can be accessed from main thread and onProgress thread, and not always protected...

File: src/Server/StaticRequestHandler.cpp (hunk 1 above)
Comment (thevar1able) [reply src/Server/StaticRequestHandler.cpp:102]: Doesn't look like it, tests are green.

**Comment by CheSema** [comment]
Just notice this PR, I also have complains about `WriteBufferFromHTTPServerResponse`
From https://github.com/ClickHouse/ClickHouse/pull/68800

> It is possible to make WriteBufferFromHTTPServerResponse and after that call HTTPServerResponse::send which creats one more WriteBufferFromPocoSocket over the same socket. There could be a race. It would be much better if WriteBufferFromHTTPServerResponse reuses HTTPServerResponse::send.

**Pull request merged**